applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h

```diff
@@ -34,35 +34,225 @@ namespace MPMSearchElementUtility
 
-    /**
-     * @brief Search element connectivity for each particle
-     * @details A search is performed to know in which grid element the material point falls.
-     * If one or more material points fall in the grid element, the grid element is
-     * set to be active and its connectivity is associated to the material point
-     * element.
-     * STEPS:
-     * 1) All the elements are set to be INACTIVE
-     * 2) A searching is performed and the grid elements which contain at least a MP are set to be ACTIVE
-     *
-     */
-    template<std::size_t TDimension>
-    void SearchElement(ModelPart& rBackgroundGridModelPart, ModelPart& rMPMModelPart, const std::size_t MaxNumberOfResults,
-        const double Tolerance)
+    typedef typename ModelPart::GeometryType GeometryType;
+
+    inline double CrossProductDet2D(array_1d<double, 3> VectorA, array_1d<double, 3> VectorB)
     {
-        const ProcessInfo& r_process_info = rBackgroundGridModelPart.GetProcessInfo();
-        const bool is_explicit = r_process_info.GetValue(IS_EXPLICIT);
+        return (VectorA[0] * VectorB[1] - VectorB[0] * VectorA[1]);
+    }
 
-        // Reset elements to inactive
-        #pragma omp parallel for
-        for (int i = 0; i < static_cast<int>(rBackgroundGridModelPart.Elements().size()); ++i) {
-            auto element_itr = rBackgroundGridModelPart.Elements().begin() + i;
-            auto& r_geometry = element_itr->GetGeometry();
-            element_itr->Reset(ACTIVE);
+    inline bool CheckIsInside(const GeometryType& rGeom, array_1d<double, 3>& LocalCoords, const array_1d<double, 3>& Coords, const double Tolerance)
+    {
+        bool is_inside = true;
+        if (rGeom.Dimension() == 2)
+        {
+            is_inside = true;
+            // Do walk around method
+            Vector cross_products(rGeom.PointsNumber());
+            for (size_t i = 0; i < rGeom.PointsNumber(); ++i)
+            {
+                if (rGeom.Points()[i].Coordinates()[2] != 0.0) {
+                    return rGeom.IsInside(Coords, LocalCoords, Tolerance);
+                    break;
+                }
+                cross_products[i] = CrossProductDet2D(Coords - rGeom.Points()[i].Coordinates(),
+                    rGeom.Points()[(i+1)% rGeom.PointsNumber()].Coordinates()- rGeom.Points()[i].Coordinates());
+            }
+            for (size_t i = 1; i < cross_products.size(); ++i)
+            {
+                if (cross_products[i] * cross_products[0] < 0.0)
+                {
+                    is_inside = false;
+                    break;
+                }
+            }
 
-            for (IndexType j = 0; j < r_geometry.PointsNumber(); ++j)
-                r_geometry[j].Reset(ACTIVE);
+        }
+
+        if (is_inside) return rGeom.IsInside(Coords, LocalCoords, Tolerance);
+
+        return false;
+    }
+
+
+    inline void ConstructNeighbourRelations(GeometryType& rGeom, const ModelPart& rBackgroundGridModelPart)
```

Comment:
cannot you use the process from the core for this?

Reply:
which one are you referring to?

Reply:
one of those:
file:///home/philippb/Pictures/Screenshot%20from%202020-06-17%2012-05-22.png
![image](https://user-images.githubusercontent.com/25484702/84885217-0479e580-b093-11ea-984e-c954153d1c88.png)

maybe @RiccardoRossi has a better overview of which one to use

Reply:
the ones in the core are either referring to elements/ conditions or nodes. This one is for geometries. Maybe we need to consider extending the core processes to geometries.

Reply:
Hey, we had a deeper look at the processes and it seems that they're not suitable at the moment. We would require deeper adaption which would involve certain add ons in the core which we would like to postpone to future PRs.